### PR TITLE
🧪🐛♻️ Google Analytics v4 Measurement Protocol: tests, bugfixes, changes, etc

### PR DESCRIPTION
**Pull request by nevans**
This PR starts on the foundation of both #98 and #100.

_I know this may be too many changes for a single PR.  And it includes some rather large changes to the API.  Sorry about that.  I can chop it up into smaller PRs or attempt to stick closer to the original API, if you would prefer that._

-----

Add `api_secret` and `client_id` requirement to Tracker.  When they are nil, NoopTracker will be used instead.

Requiring `client_id` is a big change, but I believe it is necessary for Google's v4 API.  Because access to standard analytics `client_id` is at least partially a matter of end-user consent, I believe it would be wrong for the library to automatically generate a missing `client_id`.  If users of the gem think that it is reasonable and permissible to generate a missing `client_id`, they should manage that themselves.

Other general updates:
* Added a few basic tests for V4::Tracker and V4::Event.
* Some changes to event method names to use nomenclature more similar to the Measurement API v4 documentation.
* Experimental support for auto-generating `timestamp_micros`.

Updates to Event:
* Adds `Event.param` DSL, backward-compatible with `FIELDS`.  This is mostly unused at the moment, but it's a step towards changing the `Event` module into a generic superclass.
* Return custom parameters from `#as_json`, in addition to the standard recommended parameters.
* Recursively encode `#as_json` parameter values, eg: `items: [...]`.
* Correctly encode Item, without requiring ActiveSupport.
* Add `#engagement_time_msec` and `#session_id` as recommended parameters.
* Add `#update`, for adding and updating parameters.
* Add `#apply_defaults`, for adding missing parameters to an event.

Updates to Tracker
* Add a bang-method version of each dynamically generated event type method, to immediately post all enqueued events.
* Add `tracker << event` as an alternative form of `#add`.
* Update `#add` to accept simple String, Symbol, and Hash events.
* Update `#add` to accept already constructed Event objects.
* Update `#add` to reverse merge with `#default_event_params`.
* Update `#post` visibility to be public.
* Update `#post` to send in batches of 25 events, returning an array.
* Update `#post` to clear `#events` as they are posted.
* Add `#default_event_params`, which are applied to each event.
* Add basic support for `#user_id` and `#user_properties`.
* Convert NoopTracker to subclass Tracker.
* Clear `#events` after successfully calling `#track`.

### CODE REVIEW

**Comment by tpitale** [comment]
These are all great changes. At this point, I'd rather not add _anything_ that isn't getting us closer to a working V4.

I appreciate all the work, and will definitely be open to these changes in the future (especially if, as you said, it's in smaller chunks).

I have limited time and focus for this, so I'm gonna insist for a little bit on just V4 changes.

**Comment by nevans** [comment]
@tpitale thanks.  Building on top of #100 makes it less clear, but _almost_ everything in this branch was either a minor `rubocop -a` whitespace change (that _should_ have been in #100) or in the `lib/staccato/v4**/*` files.

Please protect your limited time and focus as best you see fit.  I'm perfectly okay with you sitting on this PR for several weeks (or months).  That'll give me more time to test it in production, and iron out any further issues.  :)

**Comment by tpitale** [comment]
Yeah, let me merge your other PR and hopefully I can see the diff more easily.

**Comment by nevans** [comment]
👍🏻

You can also see https://github.com/tpitale/staccato/pull/101/files/37544fefa4ad823ec3d098e1ee98eb75df797326..dcf5c02ba22a65e0b2fcf987e6b27f327c9e38b7 for the diff _without_ the changes in #100.

**Comment by nevans** [comment]
@tpitale sorry, I missed that you had merged #100.  Thanks!  I rebased this branch on top of #98.

**Comment by asgeo1** [comment]
Hey guys, just letting you know I've been testing this PR, and it's been working for me.

So thanks for that, awesome work!

I would vote to have this PR merged into upstream repo, because it adds the important ability to send custom events, which the original PR did not have.

**Comment by tpitale** [comment]
The only reason I hesitate at this point is the extra changes that I'm not sure are necessary. I'm happy to merge those separately.

At this point I'll either merge this or pull in the necessary changes to my original PR.

File: lib/staccato/v4/tracker.rb [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+module Staccato::V4
+  # The `Tracker` class has methods to create all `Hit` types
+  #   using the measurement_id, api_secret, and client_id.
+  #
+  # @author Tony Pitale
+  class Tracker
+    extend Staccato::V4::AdapterDefaults
+
+    attr_reader :measurement_id, :api_secret
```
Comment (nevans) [inline lib/staccato/v4/tracker.rb:14]: I'm not sure how this duplicate snuck in. 
 Perhaps a rebase mistake?

```suggestion
```

File: lib/staccato/v4/tracker.rb [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+require_relative 'event' # must load first, to dynamically define event methods
+
+module Staccato::V4
+  # The `Tracker` class has methods to create all `Hit` types
+  #   using the measurement_id, api_secret, and client_id.
+  #
+  # @author Tony Pitale
+  class Tracker
+    extend Staccato::V4::AdapterDefaults
+
```
Comment (nevans) [inline lib/staccato/v4/tracker.rb:13]: Oops.  I forgot to delete this line.  That's what I get for committing a quick fix from GitHub's web UI on my phone. 😉 

```suggestion
```

**Comment by tpitale** [comment]
I think this closed because I merged the other PR. Reopening … if I can.

**Comment by tpitale** [comment]
I can't seem to re-open 🙃

**Comment by nevans** [comment]
@tpitale Yeah, I'm not sure why but sometimes GitHub automatically updates PRs based on other PR branch to be based on the main branch when the prereq is merged... and other times it does this!  I can re-create.

**Comment by tpitale** [comment]
Okay, @nevans I really like a lot of what you did here. After reviewing, I'd love to see a PR for separate additions:

1. Custom events (I think having explicit support for classes including Event, and a Hash with name/params, I'm not sure if we need to ALSO have a string/symbol, with params too. I think one or the other to start.) I think this is worth it's own PR.
2. Event defaults seems useful
3. Batch size limits, also a great addition

This will make it easier to review and merge. I appreciate all the work you've put into this, I'm sorry to ask you to do some more.

**Comment by tpitale** [comment]
@nevans Some general thoughts on the PR:

* Let's not add `attr_reader` if it's not going to be read outside the class, this exposes a public API which we may or may not be able to maintain in the future
* The client_id is not required. If it's not passed in, it'll generate a random uuid each time to fill in the blank
* Let's hold off on the DSL for the moment for `Event.param` I like where this is going, but I'd love to chat about it outside the context of V4.

**Pull request closed without merging**